compiler-plugin/src/main/kotlin/arrow/meta/quotes/MetaExtensions.kt

```diff
@@ -0,0 +1,315 @@
+import org.jetbrains.kotlin.psi.KtBlockExpression
+import org.jetbrains.kotlin.psi.KtCatchClause
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtDestructuringDeclaration
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtFinallySection
+import org.jetbrains.kotlin.psi.KtFunctionLiteral
+import org.jetbrains.kotlin.psi.KtIfExpression
+import org.jetbrains.kotlin.psi.KtImportDirective
+import org.jetbrains.kotlin.psi.KtLambdaExpression
+import org.jetbrains.kotlin.psi.KtModifierList
+import org.jetbrains.kotlin.psi.KtNamedFunction
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.kotlin.psi.KtParameterList
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtReturnExpression
+import org.jetbrains.kotlin.psi.KtThrowExpression
+import org.jetbrains.kotlin.psi.KtTryExpression
+import org.jetbrains.kotlin.psi.KtTypeAlias
+import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.kotlin.psi.KtValueArgument
+import org.jetbrains.kotlin.psi.KtWhenCondition
+
+/**
+ * ```
+ * """ $`@annotations` $expression """.annotatedExpression
+ * ```
+ */
+fun Meta.annotatedExpression(
+  match: KtAnnotatedExpression.() -> Boolean,
+  map: AnnotatedExpression.(KtAnnotatedExpression) -> Transform<KtAnnotatedExpression>
+): ExtensionPhase =
+  quote(match, map) { AnnotatedExpression(it) }
+
+/**
+ * ```
+ * """ $left $operationReference $right """".binaryExpression
+ * ```
+ */
+fun Meta.binaryExpression(
+  match: KtBinaryExpression.() -> Boolean,
+  map: BinaryExpression.(KtBinaryExpression) -> Transform<KtBinaryExpression>
+): ExtensionPhase =
+  quote(match, map) { BinaryExpression(it) }
+
+/**
+ * ```
+ * """ { $statements } """.block
+ * ```
+ */
+fun Meta.blockExpression(
+  match: KtBlockExpression.() -> Boolean,
+  map: BlockExpression.(KtBlockExpression) -> Transform<KtBlockExpression>
+): ExtensionPhase =
+  quote(match, map) { BlockExpression(it) }
+
+/**
+ * ```
+ * """ catch $`(parameter)` $`{ body }` """.catchClause
+ * ```
+ */
+fun Meta.catchClause(
+  match: KtCatchClause.() -> Boolean,
+  map: CatchClause.(KtCatchClause) -> Transform<KtCatchClause>
+): ExtensionPhase =
+  quote(match, map) { CatchClause(it) }
+
+/**
+ * ```
+ *  """|$`@annotations` $kind $name $`(typeParameters)` $`(params)` : $supertypes"} { $body } """.`class`
+ *  ```
+ */
+fun Meta.`class`(
+  match: KtClass.() -> Boolean,
+  map: ClassDeclaration.(KtClass) -> Transform<KtClass>
+): ExtensionPhase =
+  quote(match, map) { ClassDeclaration(it) }
+
+/**
+ * ```
+ * """ $`(vars)` = $initializer """.destructuringDeclaration
+ * ```
+ */
+fun Meta.destructuringDeclaration(
+  match: KtDestructuringDeclaration.() -> Boolean,
+  map: DestructuringDeclaration.(KtDestructuringDeclaration) -> Transform<KtDestructuringDeclaration>
+): ExtensionPhase =
+  quote(match, map) { DestructuringDeclaration(it) }
+
+/**
+ * ```
+ * """ $packageDirective $importList $classes """.file
```

Comment:
what about other declarations that are not classes? Are there scope properties for properties, functions and type aliases?

Reply:
Yes, there are - would you want multiple kinds of examples then?